packages/workspace/src/browser/workspace-service.ts

```diff
@@ -102,30 +213,112 @@ export class WorkspaceService implements FrontendApplicationContribution {
-            this.openWindow(uri, { preserveWindow });
+            await this.openWindow({ preserveWindow });
             return;
         }
         throw new Error(`Invalid workspace root URI. Expected an existing directory location. URI: ${rootUri}.`);
     }
 
     /**
-     * Clears current workspace root and reloads window.
+     * Adds a root folder to the workspace
+     * @param uri URI of the root folder being added
      */
-    close(): void {
-        this.doClose();
+    async addRoot(uri: URI): Promise<void> {
+        const workspaceRootChange = await this.workspace;
+        const workspaceFolder = workspaceRootChange.root;
+        if (!this.opened || !workspaceFolder) {
+            throw new Error('Folder cannot be added as there is no active folder in the current workspace.');
+        }
+
+        const rootToAdd = uri.toString();
+        const valid = await this.toValidRoot(rootToAdd);
+        if (valid) {
+            if (this._root && !this._workspaceDescription.roots.find(r => r.uri === valid.uri)) {
+                const configUri = await this.getWorkspaceConfigFileUri();
+                if (configUri) {
+                    if (!await this.fileSystem.exists(configUri.toString())) {
+                        await this.fileSystem.createFile(configUri.toString());
+                    }
+                    await this.writeRootFolderConfigFile(
+                        (await this.fileSystem.getFileStat(configUri.toString()))!,
+                        [...this._workspaceDescription.roots, valid]
+                    );
+                }
+            }
+            return;
+        }
+        throw new Error(`Invalid workspace root URI. Expected an existing directory location. URI: ${rootToAdd}.`);
+    }
+
+    /**
+     * Removes folder(s) from workspace.
+     */
+    async removeFolders(uris: URI[]): Promise<void> {
+        if (!this.opened) {
+            throw new Error('Folder cannot be removed as there is no active folder in the current workspace.');
+        }
+        const validUris: string[] = [];
+        for (const uri of uris.map(u => u.toString())) {
+            const valid = await this.toValidRoot(uri);
+            if (valid) {
+                validUris.push(uri);
+            }
+        }
+
+        if (validUris.length > 0 && this._root) {
+            const messageContainer = document.createElement('div');
+            messageContainer.textContent = 'Remove the following folders from workspace? (note: nothing will be erased from disk)';
+            const list = document.createElement('ul');
+            list.style.listStyleType = 'none';
+            validUris.forEach(u => {
+                const listItem = document.createElement('li');
+                listItem.textContent = uris.find(uri => uri.toString() === u)!.displayName;
+                list.appendChild(listItem);
+            });
+            messageContainer.appendChild(list);
+            const dialog = new ConfirmDialog({
```

Comment:
Could we move confirmation up to `navigator-contribution` that this method simply changes the file?